### PR TITLE
Adding Solar diagnostics package

### PR DESCRIPTION
**Pull request by pedro-jm**
TYPE: New feature

KEYWORDS: WRF-Solar, diagnostics

SOURCE: Timothy Juliano, Pedro Jimenez, Ju-Hye Kim (NCAR/RAL)

DESCRIPTION OF CHANGES:

Replaces PR #1047 "Adding Solar diagnostic package"

We have created a new diagnostic package for users with interests in parameters that are potentially relevant to solar applications. In the new package, we calculate: solar zenith angle, clearness index, 2-D max cloud fraction, water vapor path, liquid water path, ice water path, snow water path, total water path (liquid + ice + snow), liquid cloud effective radius, ice effective radius, snow effective radius, liquid cloud optical thickness, ice optical thickness, snow optical thickness, cloud base height, and cloud top height. Also, for liquid and ice variables, we calculate the 'total' water path, effective radius, and optical thickness, where the 'total' variables also account for subgrid (unresolved) hydrometeors. We have re-organized the radiation driver code in order to expose the calculation of unresolved hydrometeors during time steps when the radiation physics is not called. After re-organization of the radiation driver code, we ran two tests: one using the default develop code and the other using our modified code with the solar diagnostics turned off; the outputs are bit-for-bit equivalent.

All of the variables added within the solar diagnostics package are defined on the 2-D grid and also as timeseries variables in Registry/registry.solar_fields and the variables are calculated in phys/module_diag_solar.F. The solar diagnostics package is activated by setting solar_diagnostics = 1 in the WRF namelist.input file:

```
&diags
 solar_diagnostics         =  1
 /
```

LIST OF MODIFIED FILES:
M Registry/Registry.EM_COMMON
M Registry/registry.em_shared_collection
M dyn_em/module_first_rk_step_part1.F
M main/depend.common
M phys/Makefile
M phys/module_diagnostics_driver.F
M phys/module_radiation_driver.F
M share/module_check_a_mundo.F
M share/wrf_timeseries.F
M run/README.namelist
M run/README.tslist
M test/em_real/examples.namelist

LIST OF NEW FILES:
A Registry/registry.solar_fields
A phys/module_diag_solar.F

TESTS CONDUCTED: Ran a full diurnal cycle of a real case over the Pacific Northwest, USA. The simulation completed, with the output being bit-for-bit equal to the output from a simulation with the same configuration ran using an un-modified version of the code. Shown below are a few images of sample solar package variables for the domain in addition to a time series at the location indicated by a star.

swp

ts_sza

ts_wp

RELEASE NOTES:
A new diagnostic package for users with interests in parameters that are potentially relevant to solar applications is added. In the new package, The following variables are calculated: solar zenith angle, clearness index, 2-D max cloud fraction, water vapor path, liquid water path, ice water path, snow water path, total water path (liquid + ice + snow), liquid cloud effective radius, ice effective radius, snow effective radius, liquid cloud optical thickness, ice optical thickness, snow optical thickness, cloud base height, and cloud top height. Also, for liquid and ice variables, the 'total' water path, effective radius, and optical thickness are computed, where the 'total' variables also account for subgrid (unresolved) hydrometeors. The diagnostics is activated by namelist variable solar_diagnostics (=1) in namelist record @diags. (Thanks to Timothy Juliano, Pedro Jimenez, Ju-Hye Kim of NCAR/RAL)

### CODE REVIEW

File: dyn_em/module_first_rk_step_part1.F [hunk 1]
```diff
@@ -1321,10 +1289,8 @@ SUBROUTINE first_rk_step_part1 (   grid , config_flags              &
       ENDDO
       !$OMP END PARALLEL DO
      endif
-
 ! shallow cumulus parameterization
       CALL wrf_debug ( 200 , ' call shallow_cumulus_driver' )
-
```
Comment (weiwangncar) [inline dyn_em/module_first_rk_step_part1.F]: @pedro-jm Can you leave the blank lines alone?

**Comment by davegill** [comment]
@twjuliano @pedro-jm @dudhia @weiwangncar 
There seems to be LOTS of changes in the radiation driver outside the scope of the diagnostics changes. Perhaps lots of re-organizing? That really should not be part of the PR.

File: phys/module_radiation_driver.F [hunk 2]
```diff
@@ -1625,47 +1271,260 @@ SUBROUTINE radiation_driver (                                          &
...
-         DO k=kts,kte
-         DO i=its,ite
-            qc(i,k,j) = qc(i,k,j) + QCCONV(i,k,j)
-            qi(i,k,j) = qi(i,k,j) + QICONV(i,k,j)
-            ITRMX=MIN(cldfra(i,k,j),ccldfra(i,k,j))
-            ITRMN=MAX(0.,cldfra(i,k,j)+ccldfra(i,k,j)-1.)
-            cldfra(i,k,j)=cldfra(i,k,j)+ccldfra(i,k,j)-0.5*(ITRMX+ITRMN)
-         ENDDO
-         ENDDO
-         ENDDO
```
Comment (weiwangncar) [inline phys/module_radiation_driver.F]: @twjuliano @pedro-jm These were added in recent PR #921

File: phys/module_radiation_driver.F [hunk 3]
```diff
@@ -4948,9 +4729,7 @@ SUBROUTINE aer_p_int(p ,pin, levsiz, aerodt, aerod, no_src, pf, totaod,   &
          pmid(i,kk) = p(i,k,j)*0.01
       enddo
       enddo
-
    do k=1,pver
-
```
Comment (weiwangncar) [inline phys/module_radiation_driver.F]: @twjuliano @pedro-jm Please leave the blank lines as they are. Thanks.

**Comment by dudhia** [comment]
There seem to be large reorganizations in the radiation driver.

**Comment by weiwangncar** [comment]
@twjuliano @pedro-jm @dudhia As a general rule, any reorganization of the code should be done in a separate PR, not to be mixed with the actual code changes.

**Comment by pedro-jm** [comment]
The changes in the radiation driver are necessary to expose the calculation of unresolved hydrometeors to non radiation time steps (We are moving code). The diagnostics package calculates the column-integrated hydrometeor content of the unresolved hydrometeors and this is why we moved the code. Also, the unresolved hydrometeors are used by FARMS every model time step to calculate the surface irradiance. How should we move forward with this?

The removing of blank lines is a side effect of conflict resolution.  We can add them back if necessary.

**Comment by weiwangncar** [comment]
@pedro-jm OK if the moving of the code is necessary, please add some notes in your PR. Please recover some code you deleted and add the blank lines back, both in readiation driver, and module_first_rk_step_part1.F. Thanks. Because you moved code around in radiation driver, we would like to know if it has any impact in the before and after results.

**Comment by pedro-jm** [comment]
OK @twjuliano will add back the blank lines and improve the PR notes. The missing/removed code is actually there. It was just moved.

@twjuliano  did a couple of sims one with the develop branch as it was a few days ago and the other with develop + solar diag and he found bit for bit equivalent results. He is going to add this info to the PR note.

If you find any other thing let us know.

**Comment by twjuliano** [comment]
@davegill @weiwangncar @dudhia @pedro-jm I've just pushed two separate commits. The first one adds back the blank lines that were unintentionally removed in previous commits. The second one makes some minor modifications to the diag_solar code to account for unresolved hydrometeors when no only using icloud=3 in order to better align with the modifications made in the radiation driver. I hope that this PR is now good to go. Please let me know if any other changes are necessary.

Also, because @pedro-jm generated this new PR, I am not able to edit the PR note. Are you able to grant me access to edit? Thanks.

**Comment by weiwangncar** [comment]
@twjuliano Thank you for bringing back the blank lines. If you send us a few sentences about the need to re-organize the radiation driver code, we can help to add it in. 
I wonder if you could do two tests, one using the default develop code (git checkout develop), and the other using this code but with solar diag off. We would like to know if the results are identical.

File: share/wrf_timeseries.F [hunk 4]
```diff
@@ -452,11 +452,38 @@ SUBROUTINE calc_ts( grid )
...
+            grid%ts_cbaseht_tot(n,i) = grid%cbaseht_tot(ix,iy)
+            grid%ts_ctopht_tot(n,i) = grid%ctopht_tot(ix,iy)
+            grid%ts_clrnidx(n,i) = grid%clrnidx(ix,iy)
+            grid%ts_sza(n,i) = grid%sza(ix,iy)
+         END IF
 #else
          grid%ts_tsk(n,i)  = grid%nmm_tsk(ix,iy)
 #endif
          grid%ts_tslb(n,i) = grid%tslb(ix,1,iy)
-
```
Comment (davegill) [inline share/wrf_timeseries.F]: Put this line back in, too

File: share/wrf_timeseries.F [hunk 5]
```diff
@@ -483,10 +510,37 @@ SUBROUTINE calc_ts( grid )
...
+            grid%ts_ctopht(n,i) = 1.E30
+            grid%ts_cbaseht_tot(n,i) = 1.E30
+            grid%ts_ctopht_tot(n,i) = 1.E30
+            grid%ts_clrnidx(n,i) = 1.E30
+            grid%ts_sza(n,i) = 1.E30
+         END IF
 #endif
          grid%ts_tsk(n,i)  = 1.E30
          grid%ts_tslb(n,i) = 1.E30
-
```
Comment (davegill) [inline share/wrf_timeseries.F]: add this line back

File: share/wrf_timeseries.F (hunk 4 above)
Comment (twjuliano) [reply share/wrf_timeseries.F]: Done

File: share/wrf_timeseries.F (hunk 5 above)
Comment (twjuliano) [reply share/wrf_timeseries.F]: Done

**Comment by twjuliano** [comment]
@weiwangncar  I conducted these two tests yesterday and the results were identical. Unfortunately, the login nodes for Cheyenne and Casper are currently unavailable, so I am not able to access the simulation output. Hopefully I can do this tomorrow AM. Thanks.

**Comment by weiwangncar** [comment]
@twjuliano That's good to know. I am now ok with this PR. Don't forget to edit the PR message.

**Comment by twjuliano** [comment]
@weiwangncar PR message has been edited.

**Comment by davegill** [comment]
@dudhia @weiwangncar 
Folks,
This does not build with NMM
module_PHYSICS_CALLS.F is missing solar_opt

**Comment by davegill** [comment]
Closed, will use #1110 "Adding Solar diagnostics package"
This PR does not build NMM

**Pull request closed without merging**